proposals/lip-0051.md

```diff
@@ -1267,46 +1175,95 @@ unlock(address, moduleID, tokenID, amount):
         or lockedAmount(address, moduleID, tokenID) < amount):
         unlock fails
 
-    availableBalance(address, tokenID) += amount
     lockedAmount(address, moduleID, tokenID) -= amount
+    availableBalance(address, tokenID) += amount
+```
+
+#### forwardMessageFee
+
+This function is only called during the execution of a cross-chain update on the mainchain.
+```python
+forwardMessageFee(ccm):
+    fee = ccm.fee
+    feeTokenID = interoperability.getChannel(ccm.sendingChainID).messageFeeTokenID
+    if (fee < 0
+        or feeTokenID != interoperability.getChannel(ccm.receivingChainID).messageFeeTokenID):
+        forwardMessageFee fails
+
+    
+    if (feeTokenID.chainID != interoperability.getOwnChainAccount().ID
+        or escrowAmount(ccm.sendingChainID, feeTokenID.localID) < fee):
+        interoperability.terminateChain(ccm.sendingChainID)
+        return False
+
+    # transfer the fee from escrow to escrow
+    escrowAmount(ccm.sendingChainID, feeTokenID.localID) -= fee
+    escrowAmount(ccm.receivingChainID, feeTokenID.localID) += fee
 ```
 
-#### beforeExecuteCCM
+
+#### beforeApplyCCM
 
 ```python
-beforeExecuteCCM(relayerAddress, CCM):
-    fee = CCM.fee
+beforeApplyCCM(ccu, ccm):
+    relayerAddress = address derived from ccu.senderPublicKey
+    fee = ccm.fee
     if fee < 0:
-        beforeExecuteCCM fails
-
-    if interoperability.getOwnChainAccount().ID == MAINCHAIN_ID:
-        # If this chain is the mainchain, unescrow the fee
-        if escrowAmount(CCM.sendingChainID, LOCAL_ID_LSK) < fee:
-            beforeExecuteCCM fails
-        availableBalance(relayerAddress, TOKEN_ID_LSK_MAINCHAIN) += fee
-        escrowAmount(CCM.sendingChainID, LOCAL_ID_LSK) -= fee
+        beforeApplyCCM fails
+
+    feeTokenID = interoperability.getChannel(ccm.sendingChainID).messageFeeTokenID
+    # If the chain is the fee token native chain, unescrow and assign fee to relayer
+    if feeTokenID.chainID == interoperability.getOwnChainAccount().ID:
+        if escrowAmount(ccm.sendingChainID, feeTokenID.localID) < fee:
+            interoperability.terminateChain(ccm.sendingChainID)
+            return False
+
+        escrowAmount(ccm.sendingChainID, feeTokenID.localID) -= fee
+        availableBalance(relayerAddress, canonicalTokenID(feeTokenID)) += fee
+    # Else, just assign the fee to relayer
     else:
-        availableBalance(relayerAddress, TOKEN_ID_LSK) += fee    
+        availableBalance(relayerAddress, feeTokenID) += fee    
+```
+
+#### beforeRecoverCCM
+
+```python
+beforeRecoverCCM(trsSender, ccm):
+    fee = ccm.fee
+    if fee < 0:
+        beforeRecoverCCM fails
+
+    feeTokenID = interoperability.getChannel(ccm.sendingChainID).messageFeeTokenID
+    # If the chain is the fee token native chain, unescrow and assign fee to relayer
+    if feeTokenID.chainID == interoperability.getOwnChainAccount().ID:
```

Comment:
If this is not the case, no fees are credited?

Reply:
I'm not sure I understand the comment, in either case the fees are credited. I think this comment is in the same line as the previous one.
We decided to split the logic fully between fee in native token and fee non-native.

Reply:
But if this condition is not `True` no further code is executed in `beforeRecoverCCM`.

Reply:
Sorry, we were talking about different functions, I thought you were talking about `beforeApplyCCM`, where the same line exists.
In this case, we decided to leave the message fee out as it is native to a terminated chain (because the fee is always native to one of the chains of the channel). It would not be a technical problem to credit that fee, it's more a token-economics question.

Reply:
For consistency, I would think it is nicer to make this not depend on the type of token, i.e., recovering a message means reverting the state changes that were done due to emitting the message.

Reply:
Ok, will do the change.